Acceleration of Zanna-Bolton-2020 parameterization and new features required for NW2

Pperezhogin opened the pull request:
# Summary
This PR updates the code of Zanna-Bolton-2020 mesoscale parameterization and seeks three purposes:
* Include features required to run NW2 simulations.
* Accelerate the code for many CPUs. Depending on configuration, the new code is **4-100 times faster**.
* Add measurement of runtime.

# NW2
Simulations in NW2 at 1/4 resolution are now available with namelist parameters:
```
SMAG_BI_CONST = 0.06
USE_ZB2020 = True
ZB_SCALING = 2.5
ZB_SCHEME = 0
STRESS_SMOOTH_PASS = 4
ZB_KLOWER_R_DISS = 1.
```
![NW2](https://github.com/NOAA-GFDL/MOM6/assets/35234405/a4693913-043c-44ff-b036-91123235adf1)

# Code details
The code was **significantly** restructured. The full list of commits is available in a separate branch [PR_ZB_2020_NW2_and_acceleration](https://github.com/Pperezhogin/MOM6/tree/PR_ZB_2020_NW2_and_acceleration).

## Removed/new parameters
The following namelist parameters are no longer available and set to default values:
```
VG_SMOOTH_PASS=0
VG_SMOOTH_SEL=1
VG_SHARP_SEL=1
STRESS_SMOOTH_SEL=1
ZB_HYPERVISC=0
HYPVISC_GRID_DAMP
```
These parameters were controlling complicated filters options which are not currently used.

The following two parameters are introduced which control the attenuation of the subgrid model in NW2 configuration with typical values:
```
ZB_KLOWER_R_DISS=1.
ZB_KLOWER_SHEAR=1
```

## Acceleration of the code
Overall restructuring of the code with storing precomputed arrays led to acceleration ~ 4 times.

Acceleration of filters is achieved with:
* [Marching halo with grouping of non-blocking MPI exchanges](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L849-L876) (~10-100 times faster on thousands of CPUs).
* Decomposition of [2D filter](https://github.com/NOAA-GFDL/MOM6/blob/dev/gfdl/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L775-L783) into sequence of [1D filters](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L969-L975) (30% faster).
## Tests
All tests (grid, layout, nan, rotate, dim, restart, openmp) are passed.

## Regression
Regression changed twice:
*  I removed mask of [outcropped points ](https://github.com/NOAA-GFDL/MOM6/blob/dev/gfdl/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L848). The results so far **did not change qualitatively**. Support of the code for this mask is complicated and (apparently) unnecessary.
* The 2D and 1D filters differ by the round-off errors.

marshallward commented:
Thanks for putting in some of these improvements to the solver.

Before reviewing this, there's at least one error that need to be fixed.
```
  ../../../src/parameterizations/lateral/MOM_Zanna_Bolton.F90:314:29:
  
    314 |   if (CS%Stress_iter > 0 .or. CS%HPF_iter) then
        |                             1
  Error: Operands of logical operator ‘.or.’ at (1) are LOGICAL(4)/INTEGER(4)
```
In logical tests, everything has to be converted to logical (e.g. `CS%HPF_iter > 0`).

Also a good idea to start fixing up the documentation: https://github.com/NOAA-GFDL/MOM6/actions/runs/6346687254/job/17240553103?pr=484

All function arguments and derived type parameters must always be documented.

Pperezhogin commented:
Hi @marshallward, I addressed the mentioned problems with doxygen and compilation error.

marshallward commented on src/parameterizations/lateral/MOM_Zanna_Bolton.F90:
```diff
@@ -1,4 +1,5 @@
-! > Calculates Zanna and Bolton 2020 parameterization
+!> Calculates Zanna and Bolton 2020 parameterization
+!! Implemented by Perezhogin P.A. Contact: pperezhogin@gmail.com
 module MOM_Zanna_Bolton
```
Include a license statement?
```

! This file is part of MOM6. See LICENSE.md for the license.

```
You should be able to check other files for examples.

marshallward commented on src/parameterizations/lateral/MOM_Zanna_Bolton.F90:
```diff
@@ -173,726 +193,802 @@ subroutine ZB_2020_init(Time, GV, US, param_file, diag, CS, use_ZB2020)
+      else
+        Mxx(i,j) = ((CS%Txx(i,j,k)) * h(i,j,k)) * dy2h(i,j)
+        Myy(i,j) = ((CS%Tyy(i,j,k)) * h(i,j,k)) * dx2h(i,j)
+      endif
```
You may want to move the if-blocks outside of the loop for performance.

Pperezhogin replied:
As far as I know, these ifs, which evaluate the same at every loop iteration, almost do not affect the performance because of the branch prediction. They may be a bit slower for very small loops. 

I played a bit with these (and similar) if statements while preparing the PR and did not find substantial dependence of performance on the order of if and do statements.

BTW, all these loops are vectorized.

marshallward replied:
I think the result can depend on both compiler and the hardware (e.g. Intel masking instructions are about 6x faster than AMD).  But if you don't believe it's an issue then no worries.

marshallward replied:
Also FWIW, branch prediction is also going to vary across hardware.  For example, an Intel or AMD chip is probably going to have better prediction and a GPU or an A64FX core.

marshallward replied:
I just tested this in GNU 13, and inverting does appear to make a difference.

The original loop:
```
.L554:
    vmovsd  (%r9,%rax,8), %xmm0
    vmulsd  (%rbx,%rax,8), %xmm0, %xmm2
    vmulsd  (%r10,%rax,8), %xmm0, %xmm0
    vmovsd  (%r8,%rax,8), %xmm6
    vmulsd  %xmm6, %xmm2, %xmm2
    vmulsd  %xmm6, %xmm0, %xmm0
    vmulsd  (%rdx,%rax,8), %xmm2, %xmm2
    vmulsd  (%r11,%rax,8), %xmm0, %xmm0
    vmovsd  %xmm2, (%rsi,%rax,8)
    vmovsd  %xmm0, (%rcx,%rax,8)
    addq    $1, %rax
    leal    -1(%rax), %r14d
    cmpl    %r14d, %edi
    jge .L554
```
Although it's using AVX instructions, they are not parallel (mulsd vs mulpd)

Inverting the if/do blocks gave me this:
```
.L550:
    vmovupd (%r11,%rax), %ymm7
    vmulpd  (%r14,%rax), %ymm7, %ymm0
    vmulpd  (%r12,%rax), %ymm0, %ymm0
    vmovupd (%r10,%rax), %ymm7
    vmulpd  (%rbx,%rax), %ymm0, %ymm0
    vmovupd %ymm0, (%rdx,%rax)
    vmulpd  (%r14,%rax), %ymm7, %ymm0
    vmulpd  (%r12,%rax), %ymm0, %ymm0
    vmulpd  (%r9,%rax), %ymm0, %ymm0
    vmovupd %ymm0, (%rcx,%rax)
    addq    $32, %rax
    cmpq    %rax, %rsi
    jne .L550
```
This loop is using the parallel instructions.

It is possible that Intel is a bit smarter and managed to vectorize the loop, but if that is the case then it would seem to be a small amount of work to get better performance across compilers.

So I would still recommend inverting the loops, but it's up to you.

Pperezhogin replied:
Thanks for the explanations!

I updated these loops. Also I removed ifs from the [loop](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L501). I have another one place with ifs in [loop](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L586). In this case, I cannot remove ifs without significant duplication of the code (a separate loop for every combination of if conditions) or storing intermediate computations into 2D slices (which is likely less memory efficient).

marshallward commented on src/parameterizations/lateral/MOM_Zanna_Bolton.F90:
```diff
@@ -173,726 +193,802 @@ subroutine ZB_2020_init(Time, GV, US, param_file, diag, CS, use_ZB2020)
-                                    dx2q(I,J)  *S_12(I,J))) * &
-                      G%IareaCu(I,j)) / (h_u(I,j) + h_neglect)
+      h_u = 0.5 * (G%mask2dT(i,j)*h(i,j,k) + G%mask2dT(i+1,j)*h(i+1,j,k)) + h_neglect
+      fx(I,j,k) = - ((G%IdyCu(I,j)*(Mxx(i,j)                - &
```
Space after `-`

marshallward commented on src/parameterizations/lateral/MOM_Zanna_Bolton.F90:
```diff
@@ -937,40 +1033,44 @@ subroutine compute_energy_source(u, v, h, fx, fy, G, GV, CS)
-      enddo ; enddo
-    enddo
+    call cpu_clock_begin(CS%id_clock_source)
+      call create_group_pass(pass_KE_uv, KE_u, KE_v, G%Domain, To_North+To_East)
```
Extra indent?

(Actually it looks like the whole indentation of this function needs to be cleaned up.)

marshallward reviewed (changes requested):
Hi @Pperezhogin I'm finally getting around to this.  Switching from 2D to 3D halo updates looks like a great improvement, and overall structural changes looks good to me as well.

The others may also want to review these changes, but I have added some comments below.  Most are related to style (which, as de facto maintainer, you may have some liberty to tell me if I've gone too far :smile:).  There were a couple places which looked to me like it may improve performance.

* The `diff[uv]` copy in `Zanna_Bolton_2020` seemed unnecessary, since you are already modifying the field.  Can this simply be modified in-place?

* The `if (CS%Klower_R_diss > 0)` in `compute_stress_divergence` look like they can be moved outside of the do-loops.  You may see some improved performance if you invert the if/do blocks.

Other than that, it looks good to me.

Pperezhogin commented on src/parameterizations/lateral/MOM_Zanna_Bolton.F90:
```diff
@@ -173,726 +193,802 @@ subroutine ZB_2020_init(Time, GV, US, param_file, diag, CS, use_ZB2020)
+        vort_xy(I,J,k) = CS%vort_xy(I,J,k)
+      enddo; enddo
+    enddo
+  call cpu_clock_end(CS%id_clock_filter)
```
These copies are costly. Apparently, they cost the same as 4 iterations of the filter. If the updated halo size allows me to do all filter iterations in a single matching halo step, I can copy just a 2D slice instead of the full 3D array.

However, I postpone these optimizations for future. Note that the principal possibility of using this optimization depends on the number of filter iterations and/or halo size in this module which can be increased for optimization reasons.

Pperezhogin commented:
Hi @marshallward 

I addressed all you comments. Regression did not change, tests are passed. 

Regarding optimizations, I simplified if statements in loops whenever possible. On my system, it does not affect the runtime. Also, now I update diff[u,v] in-place which makes this module 5% faster.

marshallward commented:
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/21028 :heavy_check_mark: 

Thanks for patience, @Pperezhogin
